cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java

```diff
@@ -212,67 +214,106 @@ public void initialize() throws Exception {
-      if (outputKeyClass == null) {
-        outputKeyClass = TypeChecker.getGroupKeyClass(aggregator);
-      }
-      if (outputValClass == null) {
-        outputValClass = TypeChecker.getGroupValueClass(aggregator);
-      }
-      hConf.set(GROUP_KEY_CLASS, outputKeyClass.getName());
-      hConf.set(GROUP_VAL_CLASS, outputValClass.getName());
-      // in case the classes are not a WritableComparable, but is some common type we support
-      // for example, a String or a StructuredRecord
-      WritableConversion writableConversion = WritableConversions.getConversion(outputKeyClass.getName());
-      // if the conversion is null, it means the user is using their own object.
-      if (writableConversion != null) {
-        outputKeyClass = writableConversion.getWritableClass();
-      }
-      writableConversion = WritableConversions.getConversion(outputValClass.getName());
-      if (writableConversion != null) {
-        outputValClass = writableConversion.getWritableClass();
-      }
-      // check classes here instead of letting mapreduce do it, since mapreduce throws a cryptic error
-      if (!WritableComparable.class.isAssignableFrom(outputKeyClass)) {
-        throw new IllegalArgumentException(String.format(
-          "Invalid aggregator %s. The group key class %s must implement Hadoop's WritableComparable.",
-          aggregatorName, outputKeyClass));
-      }
-      if (!Writable.class.isAssignableFrom(outputValClass)) {
-        throw new IllegalArgumentException(String.format(
-          "Invalid aggregator %s. The group value class %s must implement Hadoop's Writable.",
-          aggregatorName, outputValClass));
-      }
+      String reducerName = reducers.iterator().next().getName();
+      Class<?> outputKeyClass;
+      Class<?> outputValClass;
+      if (!phaseSpec.getPhase().getStagesOfType(BatchAggregator.PLUGIN_TYPE).isEmpty()) {
+        BatchAggregator aggregator = pluginInstantiator.newPluginInstance(reducerName);
+        MapReduceAggregatorContext aggregatorContext =
+          new MapReduceAggregatorContext(context, mrMetrics,
+                                         new DatasetContextLookupProvider(context),
+                                         reducerName, context.getRuntimeArguments());
+        aggregator.prepareRun(aggregatorContext);
+        finishers.add(aggregator, aggregatorContext);
+
+        if (aggregatorContext.getNumPartitions() != null) {
+          job.setNumReduceTasks(aggregatorContext.getNumPartitions());
+        }
+        outputKeyClass = aggregatorContext.getGroupKeyClass();
+        outputValClass = aggregatorContext.getGroupValueClass();
+
+        if (outputKeyClass == null) {
+          outputKeyClass = TypeChecker.getGroupKeyClass(aggregator);
+        }
+        if (outputValClass == null) {
+          outputValClass = TypeChecker.getGroupValueClass(aggregator);
+        }
+        hConf.set(MAP_KEY_CLASS, outputKeyClass.getName());
```

Comment:
This is redundant because even though the logic is same, the context and reduce types are different.